embulk-core/build.gradle

```diff
@@ -35,10 +41,25 @@ java {
 }
 
 dependencies {
-    // Embulk plugins have expected that embulk-core's dependencies are declared as "compile" in pom.xml.
-    // Although those dependencies would be finally removed from embulk-core's direct dependencies,
-    // those "compile" declaration would be kept so that plugin developers won't be confused unnecessarily.
-    api project(":embulk-spi")
+    // The compiled class files of "embulk-spi" are mixed into the "embulk-core" JAR file, and
+    // the "embulk-core" Maven artifact has the dependency on "embulk-spi" as "provided" in pom.xml.
+    //
+    // The artifacts are structured like this so that they can bypass the restriction of Java (9+)
+    // Module System that multiple JAR files (modules) cannot contain the same Java package classes.
+    //
+    // This mixing is not necessary when using Embulk with the executable all-in-one binary. But,
+    // it would not work well with Java 9+ in case of using Embulk with self-hosted deployment of
+    // the separated artifacts, embulk-spi, embulk-core, and else.
+    spi project(":embulk-spi")
```

Comment:
Just my question.

Does `spi` statement the Embulk specific? Or, It kinds of `compieOnly`, `api` and `implementation`?

Reply:
It's specific in `embulk-core`. It's newly defined at :
https://github.com/embulk/embulk/pull/1588/files#diff-fb7f690b97a5e617e85cc5ff9468ac657d9bdfae9f5c7174e80cbad3d9898d15R18

Reply:
Thank you for your commet. LGTM👍